src/ir/context.rs

```diff
@@ -994,12 +995,31 @@ impl BindgenContext {
     /// Assign a new generated name for each anonymous field.
     fn deanonymize_fields(&mut self) {
         let _t = self.timer("deanonymize_fields");
-        let comp_types = self.items
-            .values_mut()
-            .filter_map(|item| item.kind_mut().as_type_mut())
-            .filter_map(Type::as_comp_mut);
-        for comp_info in comp_types {
-            comp_info.deanonymize_fields();
+
+        let comp_item_ids: Vec<ItemId> = self.items
+            .iter()
+            .filter_map(|(id, item)| {
+                if let Some(ty) = item.kind().as_type() {
+                    if let Some(_comp) = ty.as_comp() {
+                        return Some(id);
+                    }
+                }
+                None
+            })
+            .cloned()
+            .collect();
+
+        for id in comp_item_ids {
+            let mut item = self.items.remove(&id).unwrap();
+
+            item.kind_mut()
+                .as_type_mut()
+                .unwrap()
+                .as_comp_mut()
+                .unwrap()
+                .deanonymize_fields(self);
+
+            self.items.insert(id, item);
```

Comment:
Why do we have to remove the item and then put it back again? Why doesn't the previous approach work anymore?

This results in two passes over the items, while the previous approach was a single pass because there was no `collect` call on the iterator, just the for loop.

I doubt it is a ton of overhead, and the code isn't so difficult to follow that I'm worried about code paths where we forget to return the item into the set, but I guess I'm just a little confused and it seems like the old approach was simpler.

Reply:
Yeah, old approach was definitely simpler! I chose this approach because Context was not needed, but now it is. And because of this borrow checker become sad (because we are borrowing `self.items`). I tried to temporary `mem::replace` whole `self.items` but it didn't work out, because we actually need items array (see call to `resolve_function` in `has_method`).

So I decided to go approach like in [`compute_bitfield_units`](https://github.com/rust-lang-nursery/rust-bindgen/pull/1058/files/b708e07064fe5e39863f2d95f610d1c6129c8171#diff-caed70b64e97d2aa8acabddd9ae8e448R982).
I thought that I should factor out this pattern out, but I completely forgot about it 😳